### ISSUE TITLE
valid ksh regex match causes invalid syntax error

### ISSUE THREAD

**Issue by bahamat** [koalaman/shellcheck#2419]
#### For bugs
- Rule Id (if any, e.g. SC1000):
    - SC1009
    - SC1073
    - SC1083
    - SC1036
    - SC1072
- My shellcheck version (`shellcheck --version` or "online"): 
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/ksh

set -o xtrace

param='172.31.2.13/24'

: Shellcheck considers this a syntax error
if [[ "${param}" == {3}({1,3}(\d).){1,3}(\d)* ]]; then
    printf '== matched\n'
else
    printf '== did not match\n'
fi

: Shellcheck says this should be correct, but is not.
if [[ "${param}" =~ {3}({1,3}(\d).){1,3}(\d)* ]]; then
    printf '=~ matched\n'
else
    printf '=~ did not match\n'
fi
```

#### Here's what shellcheck currently says:

```shell
In shellcheck_bug.sh line 7:
if [[ "${param}" == {3}({1,3}(\d).){1,3}(\d)* ]]; then
^-- SC1009 (info): The mentioned syntax error was in this if expression.
   ^-- SC1073 (error): Couldn't parse this test expression. Fix to allow more checks.
                    ^-- SC1083 (warning): This { is literal. Check expression (missing ;/\n?) or quote it.
                      ^-- SC1083 (warning): This } is literal. Check expression (missing ;/\n?) or quote it.
                       ^-- SC1036 (error): '(' is invalid here. Did you forget to escape it?
                       ^-- SC1072 (error): Expected test to end here (don't wrap commands in []/[[]]). Fix any mentioned problems and try again.

For more information:
  https://www.shellcheck.net/wiki/SC1083 -- This { is literal. Check expressi...
  https://www.shellcheck.net/wiki/SC1036 -- '(' is invalid here. Did you forg...
  https://www.shellcheck.net/wiki/SC1072 -- Expected test to end here (don't ...
```

#### Here's the output from running the script:

```shell
+ param=172.31.2.13/24
+ : Shellcheck considers this a syntax error
+ [[ 172.31.2.13/24 == {3}({1,3}(\d).){1,3}(\d)* ]]
+ printf '== matched\n'
== matched
+ : Shellcheck says this should be correct, but is not.
+ [[ 172.31.2.13/24 == ~(E){3}({1,3}(\d).){1,3}(\d)* ]]
+ printf '=~ did not match\n'
=~ did not match
```

You can see that when executed, ksh converts `=~` into `== ~(E)`, which is an extended regular expression with different behavior and the match fails.

`ksh(1)` man page includes the following:

>     string == pattern
>                   True, if string matches pattern.  Any part of pattern can
>                   be quoted to cause it to be matched as a string.  With a
>                   successful match to pattern, the .sh.match array variable
>                   contains the match and sub-pattern matches.

>     string =~ ere
>                   True if string matches the pattern ~(E)ere where ere is an
>                   extended regular expression.

Spellcheck should be able to understand this correct syntax with ksh.

FWIW, this happens at least on on macOS (`/bin/ksh`) and illumos (`/usr/bin/ksh93`) (and probably Solaris) platforms. I haven't tested this with other ksh variants (e.g., mksh, pdksh).

**Comment by koalaman** [comment]
This is indeed due to ShellCheck supporting `*(..)` and `@(..)` but not ksh specific `{n}(..)`, nor the ksh parenthesis escaping like `\d` for `[[:digit:]]`. It certainly should.
